File: code/packages/python/vm-core/src/vm_core/frame.py
"""VMFrame and RegisterFile — the per-call-frame state of vm-core.

Each function call in vm-core creates a fresh VMFrame.  The frame holds:

- ``fn``          — the IIRFunction being executed
- ``ip``          — instruction pointer (index into fn.instructions)
- ``registers``   — a flat register file (one slot per virtual variable)
- ``return_dest`` — register index in the *caller* frame where the return
                    value will be stored (None for the root frame)

The RegisterFile is a plain Python list, which gives O(1) reads and writes.
Named variables are resolved by a ``name_to_reg`` dict built once per frame
from the function's parameter list — parameters occupy the first N registers,
with indices assigned in parameter order.

Example::

    fn = IIRFunction(name="add", params=[("a", "u8"), ("b", "u8")], ...)
    # a → register 0, b → register 1
    frame = VMFrame.for_function(fn)
    frame.registers[0] = 10   # a = 10
    frame.registers[1] = 20   # b = 20
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from interpreter_ir import IIRFunction


class RegisterFile:
    """A flat array of value slots, one per virtual variable.

    The file is pre-sized to ``register_count`` slots, all initialised to 0.
    Named-variable access goes through VMFrame.resolve(); direct integer-index
    access is used by the argument-passing code.
    """

    __slots__ = ("_slots",)

    def __init__(self, count: int = 8) -> None:
        self._slots: list[Any] = [0] * count

    def __getitem__(self, idx: int) -> Any:
        return self._slots[idx]

    def __setitem__(self, idx: int, value: Any) -> None:
        self._slots[idx] = value

    def __len__(self) -> int:
        return len(self._slots)

    def reset(self) -> None:
        """Zero all slots (used by REPL snapshot rollback)."""
        for i in range(len(self._slots)):
            self._slots[i] = 0

    def snapshot(self) -> list[Any]:
        """Return a shallow copy of all slots."""
        return list(self._slots)

    def restore(self, saved: list[Any]) -> None:
        """Restore slots from a snapshot."""
        for i, v in enumerate(saved):
            self._slots[i] = v


@dataclass
class VMFrame:
    """State for a single function call.

    Parameters
    ----------
    fn:
        The function being executed.
    registers:
        The register file for this frame.
    name_to_reg:
        Maps variable names to register indices.  Populated from fn.params
        at frame creation; extended by the dispatch loop as new dest names
        appear.
    ip:
        Current instruction pointer.
    return_dest:
        Register index in the *caller* frame where the return value will be
        stored, or None for the root frame.
    """

    fn: IIRFunction
    registers: RegisterFile
    name_to_reg: dict[str, int] = field(default_factory=dict)
    ip: int = 0
    return_dest: int | None = None

    @classmethod
    def for_function(
        cls,
        fn: IIRFunction,
        return_dest: int | None = None,
    ) -> VMFrame:
        """Allocate a fresh VMFrame for ``fn``.

        Parameter names are pre-assigned to registers 0..len(params)-1.
        """
        regs = RegisterFile(max(fn.register_count, len(fn.params), 1))
        name_to_reg = {name: i for i, (name, _) in enumerate(fn.params)}
        return cls(
            fn=fn,
            registers=regs,
            name_to_reg=name_to_reg,
            return_dest=return_dest,
        )

    def resolve(self, operand: Any) -> Any:
        """Resolve an operand to its current value.

        - ``str`` → look up the variable in ``name_to_reg`` and read the register
        - anything else → return as a literal
        """
        if isinstance(operand, str):
            idx = self.name_to_reg.get(operand)
            if idx is None:
                from vm_core.errors import UndefinedVariableError
                raise UndefinedVariableError(
                    f"variable {operand!r} not defined in function {self.fn.name!r}"
                )
            return self.registers[idx]
        return operand

    def assign(self, dest: str, value: Any) -> None:
        """Assign ``value`` to the variable named ``dest``.

        Allocates a new register slot if this is the first assignment to ``dest``.
        """
        if dest not in self.name_to_reg:
            next_idx = len(self.name_to_reg)
            self.name_to_reg[dest] = next_idx
        self.registers[self.name_to_reg[dest]] = value

File: code/packages/rust/vm-core/src/frame.rs
//! [`VMFrame`] — per-call-frame state of the vm-core register interpreter.
//!
//! Each function call creates a fresh `VMFrame`.  The frame holds:
//!
//! - `fn_name`       — name of the function being executed (used to look up
//!                     the `IIRFunction` in the module on each dispatch tick)
//! - `ip`            — instruction pointer (index into the function's body)
//! - `registers`     — flat register file, one slot per named variable
//! - `name_to_reg`   — variable name → register index (built from params)
//! - `return_dest`   — register index in the **caller** frame where the return
//!                     value is stored (`None` for the root frame)
//!
//! # Why `fn_name` instead of `&IIRFunction`
//!
//! Storing a name string rather than a reference breaks the borrow conflict
//! that would otherwise arise in the dispatch loop: the loop needs
//! `&mut vm.frames` (to advance `ip`) AND `&mut vm.module` (to look up the
//! callee in `handle_call`) simultaneously.  Since they're both fields of
//! `VMCore`, holding live borrows of both through `vm` is rejected by the
//! borrow checker.  Using a name + module re-lookup on each tick is the
//! clean, Rust-idiomatic solution — the lookup is a linear scan over what
//! is typically a small slice (≤ 100 functions).
//!
//! # Example
//!
//! ```
//! use vm_core::frame::VMFrame;
//! use vm_core::value::Value;
//!
//! let mut frame = VMFrame::new("add", 8, None);
//! frame.assign("a", Value::Int(10));
//! frame.assign("b", Value::Int(20));
//! assert_eq!(frame.resolve("a"), Some(&Value::Int(10)));
//! ```

use std::collections::HashMap;
use crate::value::Value;

/// Per-call-frame state of the interpreter.
#[derive(Debug)]
pub struct VMFrame {
    /// Name of the `IIRFunction` being executed.
    pub fn_name: String,

    /// Instruction pointer — 0-based index into `fn_.instructions`.
    pub ip: usize,

    /// Flat register file: `registers[i]` holds the value of whichever
    /// variable maps to register index `i`.
    pub registers: Vec<Value>,

    /// Variable name → register index.
    ///
    /// Parameters occupy the first `params.len()` slots (assigned in order);
    /// subsequent assignments extend the map dynamically.
    pub name_to_reg: HashMap<String, usize>,

    /// Register index in the **caller** frame where the return value should
    /// be written when this frame's `ret` instruction fires.
    ///
    /// `None` for the root frame (no caller exists).
    pub return_dest: Option<usize>,
}

impl VMFrame {
    /// Create a blank frame for the named function with `register_count`
    /// pre-allocated slots, all initialised to `Value::Null`.
    pub fn new(fn_name: impl Into<String>, register_count: usize, return_dest: Option<usize>) -> Self {
        VMFrame {
            fn_name: fn_name.into(),
            ip: 0,
            registers: vec![Value::Null; register_count],
            name_to_reg: HashMap::new(),
            return_dest,
        }
    }

    /// Build a frame from a function's parameter list, mapping param names to
    /// the first N register slots in declaration order.
    pub fn for_function(
        fn_name: impl Into<String>,
        params: &[(String, String)],
        register_count: usize,
        return_dest: Option<usize>,
    ) -> Self {
        let mut frame = VMFrame::new(fn_name, register_count, return_dest);
        for (idx, (param_name, _)) in params.iter().enumerate() {
            frame.name_to_reg.insert(param_name.clone(), idx);
        }
        frame
    }

    // ------------------------------------------------------------------
    // Register access
    // ------------------------------------------------------------------

    /// Return the value of a named variable, or `None` if not in scope.
    pub fn resolve(&self, name: &str) -> Option<&Value> {
        let idx = *self.name_to_reg.get(name)?;
        self.registers.get(idx)
    }

    /// Maximum number of distinct named registers per frame.
    ///
    /// Guards against unbounded register-file growth from adversarial or
    /// malformed IIR that assigns to a unique variable name on every
    /// instruction.  Assignments beyond this limit are silently dropped (the
    /// variable resolves as `Null`) rather than allocating more heap.
    pub const MAX_FRAME_REGISTERS: usize = 65_536;

    /// Assign a value to a named variable, allocating a new register if needed.
    ///
    /// Returns silently if the frame register cap [`Self::MAX_FRAME_REGISTERS`]
    /// has been reached and `name` is a new variable.  This prevents unbounded
    /// heap growth from adversarial IIR programs that generate unique variable
    /// names in a loop.
    pub fn assign(&mut self, name: &str, value: Value) {
        let next_idx = self.name_to_reg.len();
        // Enforce register-file cap for new variables.
        if next_idx >= Self::MAX_FRAME_REGISTERS && !self.name_to_reg.contains_key(name) {
            return; // SECURITY: silently drop rather than grow without bound
        }
        let idx = *self.name_to_reg.entry(name.to_string()).or_insert(next_idx);
        // Grow register file if needed (shouldn't happen for well-formed IIR).
        if idx >= self.registers.len() {
            self.registers.resize(idx + 1, Value::Null);
        }
        self.registers[idx] = value;
    }

    /// Snapshot the current register values (used by the trace system).
    pub fn snapshot(&self) -> Vec<Value> {
        self.registers.clone()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn new_frame_all_null() {
        let frame = VMFrame::new("main", 4, None);
        assert_eq!(frame.registers.len(), 4);
        assert!(frame.registers.iter().all(|v| *v == Value::Null));
    }

    #[test]
    fn for_function_maps_params() {
        let params = vec![
            ("a".to_string(), "u8".to_string()),
            ("b".to_string(), "u8".to_string()),
        ];
        let mut frame = VMFrame::for_function("add", &params, 8, None);
        frame.registers[0] = Value::Int(10);
        frame.registers[1] = Value::Int(20);
        assert_eq!(frame.resolve("a"), Some(&Value::Int(10)));
        assert_eq!(frame.resolve("b"), Some(&Value::Int(20)));
    }

    #[test]
    fn assign_creates_new_register() {
        let mut frame = VMFrame::new("f", 8, None);
        frame.assign("x", Value::Int(99));
        assert_eq!(frame.resolve("x"), Some(&Value::Int(99)));
    }

    #[test]
    fn assign_overwrites_existing() {
        let mut frame = VMFrame::new("f", 8, None);
        frame.assign("x", Value::Int(1));
        frame.assign("x", Value::Int(2));
        assert_eq!(frame.resolve("x"), Some(&Value::Int(2)));
    }

    #[test]
    fn resolve_undefined_returns_none() {
        let frame = VMFrame::new("f", 8, None);
        assert!(frame.resolve("missing").is_none());
    }
}
